File: Redcoder/Python/93.py
""" 
swapを上手く使えば分岐を消せるのでもう少しコードが短くなりそう
"""

from collections import deque
H,W,K = map(int,input().split())
box = [];
for i in range(H):
    a = []
    b = input()
    for i in range(W):
        a.append(int(b[i]))
    box.append(a)
con1 = [];
con2 = [];

for i in range(W):
    con1.append(deque())
    con1.append(deque())
ans = 0
for ii in range(H):
    for jj in range(W):
        con1 = [];
        con2 = [];

        for k in range(W):
            con1.append(deque())
            con2.append(deque())
            
        for k in range(H):
            for l in range(W):
                if k != ii or l != jj:
                    con1[l].appendleft(box[k][l])
        ans2 = 0
        change = 0
        while True:
            
            flag = True
            count = 0
            if change % 2 == 0:
                aa = 0
                for i in range(W):
                    aa += len(con1[i])
                while (aa) > 0:
                    
                    con = []
                    for i in range(W):
                        if (len(con1[i]) > 0):
                            con.append(con1[i].popleft())
                        else:
                            con.append(-1)
                    num = [[con[0],1]]
                    for i in range(1,W):
                        if num[-1][0] == con[i]:
                            num[-1][1] += 1
                        else:
                            num.append([con[i],1])
                    bb = -1
                    for i in range(len(num)):
                        if bb < num[i][1]:
                            bb = num[i][1]
                    if bb >= K:
                        for i in range(len(num)):
                            if num[i][0] != -1 and num[i][1] >= K:
                                count += num[i][0] * num[i][1]
                                flag = False
                                num[i][0] = -1
                    add = []
                    for i in range(len(num)):
                        for j in range(num[i][1]):
                            add.append(num[i][0])
                    for i in range(W):
                        if add[i] != -1:
                            con2[i].append(add[i])
                    aa = 0
                    for i in range(W):
                        aa += len(con1[i])
                change += 1
            else:
                aa = 0
                for i in range(W):
                    aa += len(con2[i])
                while (aa) > 0:
                    con = []
                    for i in range(W):
                        if (len(con2[i]) > 0):
                            con.append(con2[i].popleft())
                        else:
                            con.append(-1)
                    num = [[con[0],1]]
                    for i in range(1,W):
                        if num[-1][0] == con[i]:
                            num[-1][1] += 1
                        else:
                            num.append([con[i],1])
                    bb = -1
                    for i in range(len(num)):
                        if bb < num[i][1]:
                            bb = num[i][1]
                    if bb >= K:
                        for i in range(len(num)):
                            if num[i][0] != -1 and num[i][1] >= K:
                                count += num[i][0] * num[i][1]
                                flag = False
                                num[i][0] = -1
                    add = []
                    for i in range(len(num)):
                        for j in range(num[i][1]):
                            add.append(num[i][0])
                    for i in range(W):
                        if add[i] != -1:
                            con1[i].append(add[i])
                    aa = 0
                    for i in range(W):
                        aa += len(con2[i])
                
                change += 1
            ans2 += pow(2,change-1) * count
            if flag == True:
                break
        ans = max(ans,ans2)
print(ans)

File: Redcoder/C++/93.cpp
#include <bits/stdc++.h>
#define rep(i, nn, n) for (int i = nn; i < (n); ++i)
#define rrep(i, nn, n) for (int i = nn; i > (n); --i)
using namespace std;
using ll = long long;
using P = pair<int, int>;
#define chmin(x, y) x = min(x, y)
#define chmax(x, y) x = max(x, y)

int H, W, K;
int remove_block(vector<int> &box)
{
    vector<P> con;
    con.push_back({box[0], 1});
    rep(i, 1, W)
    {
        if (con[con.size() - 1].first == box[i])
            ++con[con.size() - 1].second;
        else
            con.push_back({box[i], 1});
    }
    int count = 0;
    int n = 0;
    rep(i, 0, con.size())
    {
        if (con[i].second >= K)
        {
            count += con[i].first * con[i].second;
            rep(j, n, n + con[i].second)
            {
                box[j] = 0;
            }
            // break;
        }
        n += con[i].second;
    }

    return count;
}
int add_point(vector<vector<int>> &box)
{
    int count = 0;
    rep(i, 0, box.size())
    {
        count += remove_block(box[i]);
    }

    return count;
}

void down_block(vector<vector<int>> &box)
{
    int n = box.size();
    rep(i, 0, W)
    {
        stack<int> con;
        rep(j, 0, n)
        {
            if (box[j][i] != 0)
            {
                con.push(box[j][i]);
            }
        }
        int m = n - 1;
        while (!con.empty())
        {
            box[m][i] = con.top();
            con.pop();
            --m;
        }
        while (m >= 0)
        {
            box[m][i] = 0;
            --m;
        }
    }
}

int main()
{
    cin >> H >> W >> K;
    vector<vector<int>> box(H, vector<int>(W, 0));
    rep(i, 0, H)
    {
        string s;
        cin >> s;
        rep(j, 0, W)
        {
            box[i][j] = int(s[j]) - '0';
        }
    }
    ll ans = 0;
    rep(i, 0, H) rep(j, 0, W)
    {
        vector<vector<int>> box2;
        copy(box.begin(), box.end(), back_inserter(box2));
        box2[i][j] = 0;

        down_block(box2);
        // printf("\n");
        // rep(i, 0, H)
        // {
        //     rep(j, 0, W)
        //     {
        //         printf("%d ", box2[i][j]);
        //     }
        //     printf("\n");
        // }

        int n = 0;
        ll num = 0;
        while (true)
        {
            int count = add_point(box2);
            num += 1LL * count * (1 << n);
            if (count == 0)
                break;

            // 石を下に移動処理
            down_block(box2);
            ++n;
        }
        ans = max(ans, num);
    }

    cout << ans << endl;
    return 0;
}